File: 2021/ruby/day5.rb
#! /usr/bin/env ruby

input = <<INPUT
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2
INPUT

# input = File.read("../inputs/day5").strip

data = input.split("\n").map do |row|
  x, y = row.split(" -> ")
  [x, y].map { |el| el.split(",").map(&:to_i) }.flatten
end
# pp data

#T1
def to_dots1(row)
  x1, y1, x2, y2 = row

  # t1
  if x1 == x2
    any_range(y1, y2).map { |y| to_key(x1, y) }
  elsif y1 == y2
    any_range(x1, x2).map { |x| to_key(x, y1) }
  else
    nil
  end
end

#T2
def to_dots2(row)
  x1, y1, x2, y2 = row
  dx = x2 - x1
  dy = y2 - y1

  if x1 == x2
    any_range(y1, y2).map { |y| to_key(x1, y) }
  elsif y1 == y2
    any_range(x1, x2).map { |x| to_key(x, y1) }
  elsif dx.abs == dy.abs
    dsx = dsig(x1, x2)
    dsy = dsig(y1, y2)
    (0..dx.abs).map { |d| to_key(x1 + d * dsx, y1 + d * dsy) }
  else
    nil
  end
end

def any_range(a, b)
  a <= b ? (a..b).to_a : a.downto(b).to_a
end

def dsig(d1, d2)
  d1 > d2 ? -1 : 1
end

def to_key(x, y)
  [x, y].join("-").to_sym
end

def prepare_res(dots)
  dots
    .compact
    .flatten
    .group_by { |el| el }
    .filter { |k, v| v.size > 1 }
    .keys
end

d1 = data.map { |row| to_dots1(row) }
res1 = prepare_res(d1)
# pp res1

puts "res1: #{res1.count}"

d2 = data.map { |row| to_dots2(row) }
res2 = prepare_res(d2)
# pp res2

puts "res2: #{res2.count}"

File: 2021/elixir/day5.exs
defmodule Day3 do
  def parse_input(input) do
    input
    |> Enum.map(&String.trim/1)
    |> Enum.map(&parse_row/1)
  end

  def parse_row(row) do
    row
    |> String.split(" -> ")
    |> Enum.flat_map(&String.split(&1, ","))
    |> Enum.map(&String.to_integer/1)
    |> List.to_tuple()
  end

  def overlaps(lines, task \\ :t1) do
    lines
    |> Enum.map(&to_dot(&1, task))
    |> Enum.reject(&is_nil/1)
    |> List.flatten()
    |> Enum.group_by(fn el -> el end)
    |> Enum.reject(fn {_k, v} -> length(v) == 1 end)
    |> Enum.map(fn {k, _v} -> k end)
    |> Enum.uniq()
  end

  # horizontal
  # {0, 9, 5, 9}
  def to_dot({x1, y, x2, y}, _) do
    for x <- x1..x2 do
      {x, y}
    end
  end

  # vertical
  # {7, 0, 7, 4}
  def to_dot({x, y1, x, y2}, _) do
    for y <- y1..y2 do
      {x, y}
    end
  end

  # diagonal
  # {1, 1, 3, 3}
  # {8, 0, 0, 8}
  # {6, 4, 2, 0}
  # T1 - remove this clause
  def to_dot({x1, y1, x2, y2}, :t2) do
    {deltax, deltay} = {x2 - x1, y2 - y1}
    da = abs(deltax)

    if da == abs(deltay) do
      {dsx, dsy} = {dsig(x1, x2), dsig(y1, y2)}

      for d <- 0..da do
        {x1 + d * dsx, y1 + d * dsy}
      end
    end
  end

  def to_dot(_, _), do: nil

  def dsig(d1, d2) do
    if d1 > d2, do: -1, else: 1
  end
end

# input = "../inputs/day5t" |> File.stream!()
input = "../inputs/day5" |> File.stream!()

# T1/2
input
|> Day3.parse_input()
|> Day3.overlaps(:t1)
# |> IO.inspect(label: "overlaps")
|> length()
|> IO.inspect(label: "res1")

# T2
input
|> Day3.parse_input()
|> Day3.overlaps(:t2)
# |> IO.inspect(label: "overlaps")
|> length()
|> IO.inspect(label: "res2")
